### DISCUSSION TITLE
Filter grouping

### DISCUSSION THREAD

**Discussion by adriangb** [geldata/gel#2129] [Q&A] [answered]
I'm trying to create a complex filter by joining simpler ones, but I'm having a bit of trouble getting what I want back.

Example data:

```
CREATE TYPE Book {
   CREATE PROPERTY genre -> str;
   CREATE PROPERTY pages -> int64;
};
CREATE TYPE Library {
   CREATE MULTI LINK books -> Book;
};
INSERT Book { genre := "fiction", pages := 1000 };
INSERT Book { genre := "nonfiction", pages := 100 };
INSERT Library { books := (SELECT Book) };
```

Now I want to find libraries that have _both_ fiction books over 500 pages and nonfiction books over 50.
In other words, the following should _not_ match:
- A single fiction book over 500 pages
- A single nonfiction book over 50 pages
- A fiction book over 500 pages and a non-fiction book under 50 pages.
- Etc.

My best attempt was the following:

```
SELECT Library
FILTER (.books.genre = "fiction" AND .books.pages > 500) AND (.books.genre = "nonfiction" AND .books.pages > 50);
```

But this returns no results, so it seems that the grouping is having no effect and so this is the same as:

```
SELECT Library
FILTER (.books.genre = "nonfiction" AND .books.pages > 500) AND ...;
```

Where the first clause matches nothing and so the second wouldn't even matter.

After some trial and error, I came up with the following, which makes sense and works, but I'm not sure if it's the best way?

```
SELECT Library
FILTER (SELECT .books FILTER .genre = "fiction").pages > 500 AND (SELECT .books FILTER .genre = "nonfiction").pages > 50;
```

**Comment by elprans** [comment, marked as answer]
@adriangb, the solution you came up with is the correct one.  The reason why your initial query returned empty results is because in the context of a single (sub-)query a longest common portion of a path (e.g. `.books`) is treated as a variable in an implicit `FOR` iteration, and a book genre cannot be both `"fiction"` and `"nonfiction"`.

To illustrate better:

```
FILTER 
    (.books.genre = "fiction" AND .books.pages > 500) 
    AND (.books.genre = "nonfiction" AND .books.pages > 50)
```

is interpreted as:

```
FILTER any(
    FOR book IN {.books}  # for each associated book
    UNION (
        book.genre = "fiction" 
        AND book.pages > 500 
        AND book.genre = "nonfiction" 
        AND book.pages > 50
    )
)
```

Your solution pushes references to `.books` into independent subqueries, where the `books` set is treated individually in each.

**Reply by adriangb** [reply]
Thank you very much for the detailed explanation.

**Comment by Mulugruntz** [comment]
@elprans 
After 3 hours of searching and twisting my query, I managed to do what I wanted, thanks to this page.

```
type SubTable {
    required property name -> str {
        constraint exclusive;
    }
}

type MyTable {
    required link sub -> SubTable;
    required property timestamp -> datetime;
    required property field1 -> uint64;
}
```

I just wanted the sum of a property, while applying a filter on another property.

```
select sum (
	(
		select (select MyTable
		filter (
			select SubTable.name = "NAME"
			and <datetime>"2014-10-31T00:00:00+00" <= .timestamp 
			and .timestamp <= <datetime>"2014-10-31T09:00:00+00")
		).field1
	)
);
```

Is the quadruple `select` necessary? Was there better? It looks awfully complex for such a simple query :-/.

**Reply by elprans** [reply]
I'm assuming you want to sum the values of `field1` of `MyTable` objects where a linked `SubTable` object has name `NAME` AND the timestamp is between the specified values.  If so, here's a simpler query:

```
WITH 
  entries := (
    SELECT MyTable
    FILTER
      .sub.name = "NAME"
      AND <datetime>"2014-10-31T00:00:00+00" <= .timestamp 
      AND .timestamp <= <datetime>"2014-10-31T09:00:00+00"
  )
SELECT sum(entries.field1);
```

**Reply by Mulugruntz** [reply]
Thanks a lot  @elprans !
I did some stuff around it to check the execution time and yours seems to be about 30% faster :-). I wonder if there is a timing tool inside edgedb already? I made my own in Python, but it's not necessarily accurate as it measures some stuff that are not only the DB processing.

Next step: changing the query to have it in batches (histogram):
With D = day, H = hour
Instead of doing it in one go, from t1 to t1+1D -> sum=10,000
Something like (pseudocode) this (if we want 1H buckets):
for n in range(24):
    t1+(n*1H) to t1+((n+1)*1H) -> sum=S_n
result <- [S_0, S_1, ..., S_23]
assert sum(result) == 10000

Easy to do in python by running X queries. But I'm wondering  if this is doable in EdgeQL (and if it makes sense?).

**Reply by Mulugruntz** [reply]
I've written this:

```
for x in {
	<datetime>"2014-10-31T00:00:00+00",
	<datetime>"2014-10-31T01:00:00+00",
	<datetime>"2014-10-31T02:00:00+00",
	<datetime>"2014-10-31T03:00:00+00"
}
UNION (
	with entries := (
		select MyTable 
		filter .sub.name = "NAME" 
		and x <= .timestamp 
		and .timestamp <= x + <duration>'1 hours' 
	) select sum(entries.field1)
);
```

It seems to work!!
And I can generate the list of times with Python. (Or is there a better way?)
However, it returns a set and not a list? Is the order guaranteed?

**Reply by elprans** [reply]
>  And I can generate the list of times with Python. (Or is there a better way?)

Yes, pass values as an array and use `array_unpack`:

```python
con.query(
    '''
    for x in {array_unpack(<array<datetime>>$my_dates)}
    UNION (
	with entries := (
	    select MyTable 
	    filter .sub.name = "NAME" 
	    and x <= .timestamp 
            and .timestamp <= x + <duration>'1 hours' 
        ) select sum(entries.field1)
    )
''', my_dates=[<list-of-python-datetimes>]);
```

> Is the order guaranteed?

In general, and like in SQL, the order of a set is not defined unless you have an explicit `ORDER BY` in your topmost `SELECT`:

```
WITH
  buckets := (
    FOR x IN {array_unpack(<array<datetime>>$my_dates)}
    UNION (
      WITH
        entries := (
          SELECT MyTable
          FILTER
            .sub.name = "NAME"
            AND x <= .timestamp
            AND .timestamp <= x + <duration>'1 hours'
        )
      SELECT (date := x, sum := sum(entries.field1))
    )
  )
SELECT buckets.sum ORDER BY buckets.date;
```
